### DISCUSSION TITLE
"Scanning" in Hex Fiend - feedback appreciated, possible feature idea

### DISCUSSION THREAD

**Discussion by Rezmason** [HexFiend/HexFiend#426] [Show and tell]
I immediately reach for Hex Fiend whenever I have a file with unknown contents, or is in an unknown format. And the first thing I do is what I call **"scanning",** which I describe below.

This is how I recovered nearly all the data in my [recent side project](https://rezmason.github.io/excel_97_egg). (Hex Fiend also made quick work of several other parts of the project!)

Does anyone else do this? Is there a better way to do this? Or is this a good basis for a new feature idea? I'd be down to try adding a new feature, if it's merited.

### Here's how to "scan" with Hex Fiend:
- open your file
- make the font nice and small
- set the views to only Line Numbers and Plain Text
- make the window tall
- _resize the window horizontally_ until any kind of pattern emerges in the window contents, which indicates the presence of something tabular with a dimension equal to the number of character columns in the Plain Text view
  - if you don't see anything, keep scrolling
  - sometimes the window has to be wider than the screen, even with a very small font drawn at the smallest supported size
  - sometimes there's a pattern, but the window size is off by a rational number, like it's 3 times too small or 5 times too big
  - resizing the window to the right width is more difficult the smaller the font glyphs are, of course

Showing beats telling. Here, watch me find a framebuffer in a memory dump:

https://github.com/HexFiend/HexFiend/assets/1041555/de146cde-da62-4236-be26-fba53610d0e7

Whoomp; there it is.

### Thoughts

This is great at finding a needle in a haystack. It's also great at finding a pögiff in a haystack. What's a pögiff? I don't know, but if it's a tabular chunk of bytes, then it's going to show up on the radar screen _even if I don't know what it is or that it's there_.

Sometimes it helps to change the text encoding, the byte grouping, and the font to see patterns that stay hidden in other configurations, but that's what those features do after all.

Obviously this use case is calling out for a new "pixel view". What's less obvious is how to add a decent set of options for that view without turning Hex Fiend into two conjoined applications. It's also worth noting that it's weird for a workflow to involve adjusting a window's width. But that's the Hex Fiend way, isn't it? That's how it's avoided adding a horizontal scroll bar.

So, I reiterate: does anyone else do this? is there a better alternative, including some other tool? and if not, does a pixel view belong in Hex Fiend, or does it merit its own application?

Thanks for your time. This program is fantastic.

**Comment by Rezmason** [comment]
Ah, I've just noticed #181 , which has several things in common with the functionality I'm envisioning. 

Supporting a pixel view would hypothetically address the readability and foreground/background color issues, while piggybacking on whatever support for byte coloring is added, including the UI.

**Comment by kainjow** [comment]
Interesting use case. What's the data format?

**Reply by Rezmason** [reply]
The one in the video? The whole file is the output straight from [ProcDump](https://en.wikipedia.org/wiki/ProcDump), literally the contents of a program's memory. The framebuffer in there is in the same format as [a BMP file's pixel storage section](https://en.wikipedia.org/wiki/BMP_file_format#Pixel_storage), and in this case each pixel is a one-byte value referencing a color in the 256-color lookup table.

But that's just in this scenario. And I only found that out after I found the framebuffer. "Scanning" finds salient data before you know its format or meaning.
